### PR TITLE
Add Dockerfile and CI

### PR DESCRIPTION
**Pull request by duskmoon314**
TLDR: Add a Dockerfile and adjust CI to build the image.

Changes:
- Add a Dockerfile to build an image based on `scratch`. This image can be used to run the app on a server without configuring `systemd` or other init systems (e.g., some commercial NAS devices may not support `systemd`).
- Add two GitHub Action jobs to build and push this image to GitHub Container Registry.
- Update several actions to the latest version to emit GitHub Action's warnings.
- Use the `gh` cli tool instead of the archived `actions/create-release` and `hub`, which was removed on 2023-10-02 (See [[All OS] hub utility will be removed on October 2](https://github.com/actions/runner-images/issues/8362))

### CODE REVIEW

**Comment by jiegec** [comment]
@z4yx I have no permission to setup ghcr registry for this repo, please have a look.

**Comment by z4yx** [comment]
Where is the entrance of creating ghcr registry?

**Comment by duskmoon314** [comment]
I think GitHub will automatically create the link `packages` in the sidebar once an image is released. The image will be listed under [https://github.com/z4yx?tab=packages](https://github.com/z4yx?tab=packages). And it should cost nothing since this is a public repo.

In this PR, the image is created only when a release happens. I have tested `push v0` to trigger a release in [my fork](https://github.com/duskmoon314/GoAuthing). I don't remember config anything, but I might have configured something before.
